cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/MinTopicLeadersPerBrokerGoal.java

```diff
@@ -0,0 +1,447 @@
+      return; // Early termination to avoid some unnecessary computation
+    }
+    for (Broker broker : clusterModel.aliveBrokers()) {
+      if (!isEligibleToHaveLeaders(broker, optimizationOptions)) {
+        continue;
+      }
+      for (String mustHaveLeaderPerBrokerTopicName : _mustHaveTopicLeadersPerBroker) {
+        int leaderCount = broker.numLeadersFor(mustHaveLeaderPerBrokerTopicName);
+        if (leaderCount < minTopicLeadersPerBroker()) {
+          throw new OptimizationFailureException(String.format("Broker %d does not have enough leader replica for topic %s. "
+              + "Minimum required per-broker leader replica count %d. Actual broker leader replica count %d",
+              broker.id(), mustHaveLeaderPerBrokerTopicName, minTopicLeadersPerBroker(), leaderCount));
+        }
+      }
+    }
+  }
+
+  /**
+   * Rebalance the given broker without violating the constraints of the current goal and optimized goals.
+   * @param broker         Broker to be balanced.
+   * @param clusterModel   The state of the cluster.
+   * @param optimizedGoals Optimized goals.
+   * @param optimizationOptions Options to take into account during optimization.
+   */
+  @Override
+  protected void rebalanceForBroker(Broker broker,
+                                    ClusterModel clusterModel,
+                                    Set<Goal> optimizedGoals,
+                                    OptimizationOptions optimizationOptions) throws OptimizationFailureException {
+    LOG.debug("balancing broker {}, optimized goals = {}", broker, optimizedGoals);
+    moveAwayOfflineReplicas(broker, clusterModel, optimizedGoals, optimizationOptions);
+    if (_mustHaveTopicLeadersPerBroker.isEmpty()) {
+      return; // Early termination to avoid some unnecessary computation
+    }
+    if (!isEligibleToHaveLeaders(broker, optimizationOptions)) {
+      return;
+    }
+    for (String topicMustHaveLeaderPerBroker : _mustHaveTopicLeadersPerBroker) {
+      maybeMoveLeaderOfTopicToBroker(topicMustHaveLeaderPerBroker, broker, clusterModel, optimizedGoals, optimizationOptions);
+    }
+  }
+
+  private void maybeMoveLeaderOfTopicToBroker(String topicMustHaveLeaderPerBroker,
+                                              Broker broker,
+                                              ClusterModel clusterModel,
+                                              Set<Goal> optimizedGoals,
+                                              OptimizationOptions optimizationOptions) throws OptimizationFailureException {
+    int topicLeaderCountOnReceiverBroker = broker.numLeadersFor(topicMustHaveLeaderPerBroker);
+    if (topicLeaderCountOnReceiverBroker >= minTopicLeadersPerBroker()) {
+      return; // This broker has enough leader replica(s) for the given topic
+    }
+    // Try to elect follower replica(s) of the interested topic on this broker to be leader
+    Set<Replica> followerReplicas = broker.trackedSortedReplicas(_replicaSortName)
```

Comment:
Set does not guarantee order preservation. Should we use a `LinkedHashSet` (or a `List`) here to preserve the order? -- e.g.

```
    Set<Replica> followerReplicas = broker.trackedSortedReplicas(_replicaSortName)
                                          .sortedReplicas(false).stream()
                                          .filter(r -> !r.isLeader() && r.topicPartition().topic().equals(topicMustHaveLeaderPerBroker))
                                          .collect(Collectors.toCollection(LinkedHashSet::new));
```

Reply:
I will go with `List`